lib/ecmascript.ts

```diff
@@ -1130,73 +1158,94 @@ function MergeLargestUnitOption<T extends Temporal.DateTimeUnit>(
+    const [property, defaultValue] = wrappedFieldRecord;
+    let value = bag[property];
     if (value !== undefined) {
       any = true;
       if (BUILTIN_CASTS.has(property)) {
-        result[property] = (BUILTIN_CASTS.get(property) as BuiltinCastFunction)(value);
-      } else {
-        result[property] = value;
+        // We just has-checked this map access, so there will definitely be a
+        // value.
+        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+        value = BUILTIN_CASTS.get(property)!(value);
       }
-    }
-  }
-  return any ? (result as B) : false;
-}
-
-export function PrepareTemporalFields<B extends AnyTemporalLikeType>(
-  bagParam: B,
-  fieldsParam: ReadonlyArray<FieldRecord<B>>
-): Required<B> {
-  // External callers are limited to specific types, but this function's
-  // implementation uses generic property types. The casts below (and at the
-  // end) convert to/from generic records.
-  const bag = bagParam as Record<PrimitivePropertyNames & keyof B, string | number | undefined>;
-  const fields = fieldsParam as ReadonlyArray<FieldRecord<typeof bag>>;
-  const result = {} as typeof bag;
-  let any = false;
-  for (const fieldRecord of fields) {
-    const [property, defaultValue] = fieldRecord;
-    let value = bag[property];
-    if (value === undefined) {
-      if (fieldRecord.length === 1) {
+      result[property] = value as string | number | undefined;
+    } else if (completeness !== 'partial') {
+      if (wrappedFieldRecord.length === 1) {
         throw new TypeError(`required property '${property}' missing or undefined`);
       }
-      value = defaultValue;
-    } else {
-      any = true;
-      if (BUILTIN_CASTS.has(property as PrimitivePropertyNames)) {
-        value = (BUILTIN_CASTS.get(property) as BuiltinCastFunction)(value);
-      }
+      value = defaultValue as unknown as typeof value;
+      result[property] = value as string | number | undefined;
     }
-    result[property] = value;
   }
-  if (!any) {
-    throw new TypeError('no supported properties found');
+  if (completeness === 'partial' && !any) {
+    throw new TypeError(emptySourceErrorMessage);
   }
-  if (
-    ((result as { era: unknown })['era'] === undefined) !==
-    ((result as { eraYear: unknown })['eraYear'] === undefined)
-  ) {
+  if ((result.era === undefined) !== (result.eraYear === undefined)) {
     throw new RangeError("properties 'era' and 'eraYear' must be provided together");
   }
-  return result as unknown as Required<B>;
+  return result as unknown as Owner<Exclude<FieldKeys, undefined>>;
 }
 
 // field access in the following operations is intentionally alphabetical
 export function ToTemporalDateFields(
   bag: Temporal.PlainDateLike,
-  fieldNames: readonly (keyof Temporal.PlainDateLike)[]
-) {
-  const entries: [keyof Temporal.PlainDateLike, 0 | undefined][] = [
+  fieldNames: readonly (keyof Temporal.PlainDateLike & PrimitivePropertyNames)[]
```

Comment:
The `keyof Temporal.X & PrimitivePropertyNames` pattern is used in a lot of different places in the code. Should it be abstracted into a generic type e.g. `PrimitivePropsOnly<T>`?  Or would doing this break IDE autocomplete and/or TS compilation due to limitations in TS's understanding of generics?

Reply:
I abstracted it, and it both compiles and autocomplete works correctly still. Thanks for the suggestion - this was useful to solve the comment r.e. calendars as well.